io_port.c

```diff
@@ -110,15 +125,52 @@ io_port_deregister(void)
 void
 io_port_write(uint64_t addr, uint64_t size, uint32_t val)
 {
+    uint64_t port_index;
+    uint64_t alignment;
+
     assert(io_port.enable && io_port.addr <= addr
            && addr < (io_port.addr + io_port.size));
 
     addr -= io_port.addr;
 
-    if (size == 4)
-        io_port.writel(addr, val);
+    alignment = addr % sizeof(int32_t);
+
+    if (size + alignment <= sizeof(int32_t)) {
+        port_index = addr / sizeof(int32_t);
+        if (port_index < NUMBER_OF_PORTS && io_port.writel[port_index] != NULL)
+           io_port.writel[port_index](addr - port_index * sizeof(int32_t), size, val);
+        else
+           DBG("Write to offset %" PRIu64 "not supported.\n", addr);
```

Comment:
should this return an error instead of silently succeeding? Same question about the other else below

Reply:
The guest interface to IO ports is just an asm instruction like `outl`. It's not making a hypercall or something so there isn't much scope for returning an error.